Proposal: simpler LayerStack and DerivedLayers

simbilod opened the pull request:
The current way to deal with "derived layers" (i.e. physical levels on the chip that do not map 1:1 to GDS design layers) is very convoluted. It requires defining etch and grow levels, with flags for what is etched into what, which borders on process emulation. This also does not generalize well to other common mapping between GDS layers and physical elements on the chip, for instance marker layers modifying shared process steps.

I propose instead, at the gdsfactory level, to only consider the final fabricated levels of the chip. These layers should be easily defined from **arbitrary combinations of GDS layers**. An example could look something like this:

```
    layer1 = GDSLayer(layer=(1, 0))
    layer2 = GDSLayer(layer=(2, 0))

    ls = LayerStack(
        layers={
            "layerlevel_layer1": LayerLevel(layer=layer1, thickness=10, zmin=0),
            "layerlevel_layer2": LayerLevel(layer=layer2, thickness=10, zmin=10),
            "layerlevel_and_layer": LayerLevel(
                layer=layer1 & layer2,
                thickness=10,
                zmin=0,
                derived_layer=GDSLayer(layer=(3, 0)),
            ),
            "layerlevel_xor_layer": LayerLevel(
                layer=layer1 ^ layer2,
                thickness=10,
                zmin=0,
                derived_layer=GDSLayer(layer=(4, 0)),
            ),
            "layerlevel_not_layer": LayerLevel(
                layer=layer1 - layer2,
                thickness=10,
                zmin=0,
                derived_layer=GDSLayer(layer=(5, 0)),
            ),
            "layerlevel_or_layer": LayerLevel(
                layer=layer1 | layer2,
                thickness=10,
                zmin=0,
                derived_layer=GDSLayer(layer=(6, 0)),
            ),
            "layerlevel_composed_layer": LayerLevel(
                layer=layer1 - (layer1 ^ layer2),
                thickness=10,
                zmin=0,
                derived_layer=GDSLayer(layer=(7, 0)),
            ),
        }
    )

    # Test with simple component
    import gdsfactory as gf

    c = gf.Component()

    rect1 = c << gf.components.rectangle(size=(10, 10), layer=(1, 0))
    rect2 = c << gf.components.rectangle(size=(10, 10), layer=(2, 0))
    rect2.dmove((5, 5))
    c.show()
```
![image](https://github.com/gdsfactory/gdsfactory/assets/46427609/c624d6e0-0e96-4669-9b5a-3a98ab32bff9)

Now get the component with derived layers according to the stack above:
```
    c = get_component_with_derived_layers(c, ls)
    c.show()
```

(3,0) is the intersection:

![image](https://github.com/gdsfactory/gdsfactory/assets/46427609/655db5a2-fb32-4f73-bce2-341bb05d163e)

(4,0) is the exclusive or:

![image](https://github.com/gdsfactory/gdsfactory/assets/46427609/a0cd08e3-fb42-4d25-83a5-e144a9c06e0a)

(5,0) is a substraction:

![image](https://github.com/gdsfactory/gdsfactory/assets/46427609/537db7a9-e3cb-4667-813d-ce84af9079da)

(6,0) is union:

![image](https://github.com/gdsfactory/gdsfactory/assets/46427609/69b66387-ae0c-4145-a2dd-03af67f54ee7)

We can also compose the operations; (7,0) substracts the X0R from layer 1:

![image](https://github.com/gdsfactory/gdsfactory/assets/46427609/c1505374-3630-458a-85c6-8e8a86cc601c)


We can revisit process emulation and/or simulation in gplugins.

If you think this is a good idea I can make tests and fix visualization scripts

simbilod commented:
I'm also removing all the clutter from as discussed in #2556

joamatab commented:
looks great, notice there are some conflicts

joamatab commented on gdsfactory/technology/layer_stack.py:
```diff
@@ -8,65 +8,212 @@
     from gdsfactory.technology import LayerViews
 
 
+class GDSLayer(BaseModel):
```
How about calling it LogicalLayer?

simbilod replied:
done!

joamatab reviewed (commented):
looks great!

simbilod commented:
get_klayout_3d_script needs fixing, but it should be easy to convert the boolean operations in the layerstack to klayout script booleans

simbilod commented:
there's issues with Pydantic to_json, will check again later

simbilod commented:
> there's issues with Pydantic to_json, will check again later

the pydantic serialization error is fixable by allowing layers to be int, is this acceptable behavior @simbilod @sebastian-goeldi ?

simbilod commented on gdsfactory/technology/layer_stack.py:
```diff
@@ -14,59 +14,234 @@
+class LogicalLayer(BaseModel):
+    """GDS design layer."""
+
+    layer: tuple[int, int] | kf.LayerEnum | int
```
here

simbilod commented on gdsfactory/technology/layer_stack.py:
```diff
@@ -14,59 +14,234 @@
+    """
+
+    layer1: LogicalLayer | DerivedLayer | int
+    layer2: LogicalLayer | DerivedLayer | int
```
and here

sebastian-goeldi commented:
> > there's issues with Pydantic to_json, will check again later
> 
> the pydantic serialization error is fixable by allowing layers to be int, is this acceptable behavior @joamatab @sebastian-goeldi ?

The problem with `int` is, it's not a good identifier, even `LayerEnum` is questionable in this case. Anything that should live with layers outside or directly tied info, should use something more akin to `gf.kdb.LayerInfo`, so tuple with layer datatype or a string with layer name are better choices.

An int should be stricly used for layout purposes (it's even not so good to pass it through function calls tbh as they will be serialized as ints)

To get tuple from LayerEnum:
`tuple(LAYER.WG)` for example

To get name:
`LAYER.WG.name`

To get layerenum from tuple:
`LAYER(gf.kcl.layer(*layer_tuple))` (or `layer_tuple[0],layer_tuple[1]` instead o `*...` of course.

Get LayerEnum by name:
`LAYER[layer_name]`

joamatab commented:
when will this PR be ready to review? we plan to release gdsfactory8 in the coming days

simbilod commented:
Of we can disable export KLayout 3D view, it can be done soon. Otherwise it will take longer

joamatab commented:
yes, we can fix the export to klayout3D later

simbilod commented:
I've implemented @HelgeGehring 's suggestions, it still works exactly as I showed above

@joamatab this is ready to review. I've added a NotImplementedError on the klayout 3d script for now. The only sketchy thing is the int type allowed for layer object to fix pydantic issues. Also the commit history is a bit messy, we can rebase against main if you want.

simbilod commented:
Some export tests are still failing, probably just need to call `layer.layer` in some places. I can look tomorrow

joamatab commented:
Sounds good, 

will this work with the plugins? like gmsh?

im having lots of issues with the gdsfactory8 brach in gplugins

i think it would be great to use klayout instead of shapely there if possible

simbilod commented:
> Sounds good,
> 
> will this work with the plugins? like gmsh?
> 
> im having lots of issues with the gdsfactory8 brach in gplugins
> 
> i think it would be great to use klayout instead of shapely there if possible



> Sounds good,
> 
> will this work with the plugins? like gmsh?
> 
> im having lots of issues with the gdsfactory8 brach in gplugins
> 
> i think it would be great to use klayout instead of shapely there if possible

The meshing should mostly be fine, it only relies on get_component or get_component_with_derived_layers, so if those work it will be OK. We might need to add a layer.layer here and there at most.

sebastian-goeldi commented:
I think for layers in general, I think we should just pass them as strings (i.e. layer name).

The longer the more I think it's easier to do it that way. Getting them can be as easy as `kcl.get_layer`. This also allows for easier and more meaningful (de)serialization of layers as the index itself has next to no meaning (except during construction)
